### ISSUE TITLE
id3v23 not working when Zero plugin is used (Beets v 1.3.17)

### ISSUE THREAD

**Issue by dannn-o** [beetbox/beets#1903]
Even with "**id3v23: yes**" coded in the Beets config.yaml file, tags are getting stored with the ID3 v2.4 formatting (as confirmed with MusicBrainz Picard and Puddletag) whenever the Zero plugin is also used.  When the Zero plug is removed from the config.yaml file, the "**id3v23: yes**" entry is honored and the tags are stored in ID3 v2.3 format.  Version of Beets is 1.3.17.

**Comment by sampsyo** [comment]
Oddly, I can't reproduce this here. I tried taking a file with ID3v2.4 tags:

```sh
$ beet ls -p
/Users/asampson/code/beets/_etc/testmusic/m83/01 Intro.mp3
$ file _etc/testmusic/m83/test.mp3
_etc/testmusic/m83/test.mp3: Audio file with ID3 version 2.4.0, contains: MPEG ADTS, layer III, v1, 128 kbps, 44.1 kHz, JntStereo
```

and then turning on both the `zero` plugin and the `id3v23` option. Then:

```sh
$ beet write
[...]
$ file _etc/testmusic/m83/test.mp3
_etc/testmusic/m83/01 Intro.mp3: Audio file with ID3 version 2.3.0, contains: MPEG ADTS, layer III, v1, 128 kbps, 44.1 kHz, JntStereo
```

showing the ID3 version is as expected.

I'm not sure what else to try. Is there a specific set of steps you can demonstrate so I can reproduce the problem?

**Comment by dannn-o** [comment]
Well, the good news (I guess) is that I seem to be able to reproduce the issue fairly easily.

I picked an album I recently imported, and copied it back from my FreeNAS box to my local Music folder:
`/home/dan/Music/Takin' It to the Streets`

Then, checked the ID3 version with the file command to mimic your steps:

```
$ file "/home/dan/Music/Takin' It to the Streets/01 Wheels of Fortune.mp3"
 /home/dan/Music/Takin' It to the Streets/01 Wheels of Fortune.mp3: Audio file with ID3 version 2.3.0, contains: MPEG ADTS, layer III, v1, 192 kbps, 44.1 kHz, Stereo
```
Next, I performed the import:
```
$ beet import "/home/dan/Music/Takin' It to the Streets"

/home/dan/Music/Takin' It to the Streets (9 items)
Tagging:
    The Doobie Brothers - Takin' It to the Streets
URL:
    http://musicbrainz.org/release/4bdccba9-354c-4609-9c55-c6e0ff5a07a8
(Similarity: 100.0%) (12" Vinyl, 1976, US, Warner Bros. Records)
```
Then, I repeated the file command again (note new location of the file, as 
**move: yes** is coded in the config file):
```
$ file "/home/dan/Music/The Doobie Brothers/Takin' It to the Streets/01 Wheels of Fortune.mp3"
/home/dan/Music/The Doobie Brothers/Takin' It to the Streets/01 Wheels of Fortune.mp3: Audio file with ID3 version 2.4.0, contains: MPEG ADTS, layer III, v1, 192 kbps, 44.1 kHz, Stereo
```
Hoping that will help some.  Let me know if there is any other info I can help provide.

**Comment by sampsyo** [comment]
Does `beet write` have an effect?

**Comment by dannn-o** [comment]
Since I have **write: yes** coded within the configuration's Import section, would you want me to remove that first?  And then perform an Import, followed by a Write?

**Comment by sampsyo** [comment]
No need to change your config; it would just be helpful to have you follow exactly the same steps as me (above) to see if you get the same result.

**Comment by dannn-o** [comment]
It appears I had the same outcome in repeating your steps from above:
```
$ file "/home/dan/Music/The Doobie Brothers/Takin' It to the Streets/01 Wheels of Fortune.mp3"
/home/dan/Music/The Doobie Brothers/Takin' It to the Streets/01 Wheels of Fortune.mp3: Audio file with ID3 version 2.4.0, contains: MPEG ADTS, layer III, v1, 192 kbps, 44.1 kHz, Stereo
$ beet write
....
$ file "/home/dan/Music/The Doobie Brothers/Takin' It to the Streets/01 Wheels of Fortune.mp3"
/home/dan/Music/The Doobie Brothers/Takin' It to the Streets/01 Wheels of Fortune.mp3: Audio file with ID3 version 2.3.0, contains: MPEG ADTS, layer III, v1, 192 kbps, 44.1 kHz, Stereo

```
Not sure, but does that implicate the logic involved with Import that calls the Write function (as opposed to when Write is invoked directly, which appears to be working correctly)?

**Comment by jackwilsdon** [comment]
So after doing some digging, it looks like the write setting for import [calls `Item.try_write`](https://github.com/beetbox/beets/blob/abf2511002ae95d204b7ce04a6b35904e85d52dd/beets/importer.py#L682), whereas `beet write` [calls `Item.try_sync`](https://github.com/beetbox/beets/blob/abf2511002ae95d204b7ce04a6b35904e85d52dd/beets/ui/commands.py#L1560).

Now the odd thing is that `Item.try_sync` ends up calling `Item.try_write` when the first argument is True, which is the case for `beet write`.

I honestly have no idea what could be causing this, as they both seem to end up calling `Item.try_write` (and in turn `Item.write`). Could you remove the offending song[s] from your library and reimport them using `beet -vv` and post the logs? Could you do the same for `beet -vv write` too?

**Comment by dannn-o** [comment]
As Johnny Bravo used to say: "This just keeps gettin' weirder and weirder..."

I added **log: /home/dan/Music/import-log.txt** to the config file, and then reimported an album with the -vv parameter added.  The resulting output is attached in 
[import-log.txt](https://github.com/beetbox/beets/files/164289/import-log.txt)

The only line directly written to the log file was the first line.  All subsequent lines (which follow the row of dashes) I had to copy-and-paste from the terminal session (I don't know if that is a 'bug' or not, but we can leave that for later).  The 'weird' thing I encountered is that the imported MP3 files now (correctly) have tags in the ID3v2.3 format.  It seems that enabling the very verbose logging somehow masked the problem (which I confirmed by repeating the steps a second time, with the same results).

As a sanity check, I repeated the steps again, this time omitting the -vv parameter, and the problem of the tags being stored in ID3v2.4 format returned again.  That produced the final "import started" entry at the end of the log.

I went ahead and performed a beet write -vv as requested, and the log is attached:
[write-log.txt](https://github.com/beetbox/beets/files/164302/write-log.txt)
I'm guessing this might not be helpful now, since the Import function had been successful, but I'm including it for completeness.

**Comment by sampsyo** [comment]
Ugh, that is incredibly weird. I'll keep trying to reproduce the problem too and, in the mean time, let us know if you can confirm some other factor that makes the bug reliably appear or disappear.

**Comment by dannn-o** [comment]
As an IT guy, I know how weird a technical issue can get, but I'm a bit worried that no one there seems to be able to reproduce the problem.  I will try to think of some scenarios to test different config/plugin settings that might shed more light.  In case it is something unique to my setup that is causing the problem, it might be worth mentioning that I normally clear out my beets database after each successful import by using **beet remove** (I may have mentioned in the forum that I'm still transitioning from Windows to Linux, and am still managing my music collection manually until I figure things out further).  Other than that, I think my usage can be classified as 'basic'.

**Comment by dannn-o** [comment]
I finally got some time today to experiment with my config.yaml setup, and had some interesting results.  When I started, this was the configuration used:
```
directory: ~/Music
library: ~/data/musiclibrary.blb

id3v23: yes
plugins: lyrics fetchart embedart scrub zero

import:
    move: yes
    write: yes
    log: /home/dan/Music/import-log.txt

zero:
    fields: comments
    update_database: yes
```
To begin, I removed the entire **plugins:** entry to see if the contents and/or order had any impact, and my first test was to include just the Zero plugin by itself:
`plugins: zero`
After importing an album, I checked and discovered that the tags did store in id3v23 format successfully.  I then tried using several different permutations of the **plugins:** values, and these were the results (OK = id3v23, WRONG=id3v24):

- zero (OK)
- zero lyrics (OK)
- zero lyrics fetchart (OK)
- zero lyrics fetchart embedart scrub (OK)
- lyrics zero fetchart embedart scrub (OK)
- lyrics fetchart zero embedart scrub (WRONG)
- lyrics embedart scrub zero fetchart (WRONG)
- lyrics embedart zero scrub fetchart (WRONG)
- lyrics zero embedart scrub fetchart (WRONG)
- lyrics zero fetchart embedart scrub (OK)
- lyrics scrub zero fetchart embedart (WRONG)
- scrub zero lyrics fetchart embedart (WRONG)
- zero scrub lyrics fetchart embedart (OK)

I decided to stop there, and left the last entry as my current setting since it seems to work.  I was hoping to see some sort of pattern, but the only (anecdotal) conclusion seems to be that **zero** seems to work better when it's toward the front of the pack.

I'm hoping that this will shed a bit more light on the issue, and possibly help for someone on the development side to be able to recreate the problem.  BTW - I ran into a "glitch" with the lyrics plugin during this testing, but will plan to post that in the forum separately.

**Comment by sampsyo** [comment]
Wow! That's certainly strange.

My best bet currently is that it's a bad interaction with scrub. (My second best guess is embedart.) If you get a chance, could you try enabling *just* those two plugins to see if that changes anything?

**Comment by dannn-o** [comment]
Here are different combinations I tried:
- zero scrub (OK)
- scrub zero (OK)
- zero embedart (OK)
- embedart zero (OK)
- embedart scrub (OK)
- scrub embedart (OK)
- zero scrub embedart (OK)
- scrub zero embedart (OK)
- scrub embedart zero (OK)

Not sure how much this helps.  Let me know if there are any other
combos you'd like me to try.

**Comment by sampsyo** [comment]
Wow; weird. So the conclusion must be, given your previous experience, that fetchart or lyrics must also be enabled? I suppose it's reasonable to guess that fetchart would be needed to make embedart activate.

**Comment by dannn-o** [comment]
Still unsure at this point.  Looking at the 2 lists above, I thought of a few potentially interesting combos:
- scrub zero fetchart embedart (OK)
- scrub zero fetchart embedart lyrics (OK)
- scrub zero fetchart lyrics embedart (OK)
- scrub zero lyrics fetchart embedart (WRONG - in 1st list above, but ran with same results again)
- scrub lyrics fetchart embedart (OK)
- scrub lyrics fetchart embedart zero (WRONG)
- lyrics fetchart embedart zero (OK)
- fetchart embedart scrub zero (WRONG)
- fetchart embedart zero (OK)
- lyrics scrub zero (OK)
- lyrics scrub zero fetchart (OK)

At this point, though, the few hypotheses I had have been disproved, and I fear that if I continue, I'll just obfuscate things because there are too many permutations.  It is a bit interesting, though, that a combo with only 4 plugins (fetchart embedart scrub zero) reproduced the problem, and lyrics wasn't involved in that one.

**Comment by sampsyo** [comment]
Aha! I know it's *increidbly* weird that only a four-way interaction seemed to trigger this, but it actually helped me track it down. There's a point in the `scrub` plugin that was failing to re-save the file using ID3v2.3 when it needed to restore album art.

I've pushed a fix, above. Could you try it out and check whether it makes the problem go away?

**Comment by dannn-o** [comment]
Would be glad to give it a try when I get home tonight.  Any chance you can help with basic instructions on how I incorporate the fix on my computer?

**Comment by jackwilsdon** [comment]
Take a look at [how to run the latest version of beets](https://beets.readthedocs.org/en/latest/faq.html#run-the-latest-source-version-of-beets). Note it's a "bleeding edge" build, so some things may not work (although I'm running the latest commit and it's working fine for me!).

**Comment by dannn-o** [comment]
Thanks for the link.  That made it quite easy, especially since I'm using pip.  The latest source install went OK (beets version showing 1.3.18).  I ran through each of the failing combinations above, and I'm happy to report that they all worked successfully, with the ID3 tags storing in v2.3 format each time.  I appreciate the great support for such an awesome program, and I'm glad the issue ultimately got resolved.
